bluetooth: hci: spi: Fix events being discardable, RX buffer corruption

JordanYates opened the pull request:
Extract the HCI message handling out to a dedicated function to simplify
the main thread function. This also solves a bug as a side effect.

Previously `discardable` and `timeout` were never being reset after
an advertising report was received, resulting in ALL events after the
first advertising report being treated as discardable.

Don't re-use the `rxmsg` buffer in the `bt_spi_send` function. This
buffer is still used by the RX thread after releasing the SPI semaphore.
The current re-use can result in buffer corruption if the RX thread is
swapped out as a result of the `k_sem_give`.

Moving the semaphore release later can result in deadlocks due to
buffer allocation being performed while holding the semaphore, so
instead just eliminate the re-use entirely.

Commit:
bluetooth: hci: spi: extract message handling to func

Extract the HCI message handling out to a dedicated function to simplify
the main thread function. This also solves a bug as a side effect.

Previously `discardable` and `timeout` were never being reset after
an advertising report was received, resulting in ALL events after the
first advertising report being treated as discardable.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

jhedberg commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -285,17 +285,72 @@ static int bt_spi_send_aci_config_data_controller_mode(void)
 #endif /* CONFIG_BT_BLUENRG_ACI */
 
-static void bt_spi_rx_thread(void)
+static struct net_buf *bt_spi_rx_buf_construct(uint8_t *rxmsg)
```
Not completely sure about the "construct" name, since it doesn't only do that: it actually consumes the data in the case of a vendor packet.

JordanYates replied:
Sort of true, the only event `bt_spi_handle_vendor_evt` ever actually consumes is `EVT_BLUE_INITIALIZED`, everything else is passed up the stack. Originally I called it `handle`, but its not really doing that either, just constructing the `net_buf`. Happy to hear any other name suggestions.

JordanYates commented:
Current solution is subtly incorrect. While it fixes the `rxmsg` overwrite problem, it can trigger deadlocks because the RX thread can now stop TX packets going out, if blocking on `bt_buf_get_evt` in `bt_spi_rx_buf_construct`.

Working on alternate solution, likely by eliminating the buffer re-use entirely.

Commit:
bluetooth: hci: spi: avoid `rxmsg` re-use

Don't re-use the `rxmsg` buffer in the `bt_spi_send` function. This
buffer is still used by the RX thread after releasing the SPI semaphore.
The current re-use can result in buffer corruption if the RX thread is
swapped out as a result of the `k_sem_give`.

Moving the semaphore release later can result in deadlocks due to
buffer allocation being performed while holding the semaphore, so
instead just eliminate the re-use entirely.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
Buffer corruption fixed by removing re-use of `rxmsg` in `bt_spi_send`

alwa-nordic reviewed (approved):
I'm not familiar with this driver. LGTM under the assumption `spi_transceive_dt` will continue the TX operation when it runs out of RX buffer. (It will discard the RX bytes that don't fit.) And that this behavior does not depend on the SPI driver. This seems to not be documented on the API.
